File: Multimodal Analysis_Tsirakou.py
# The code below aims to simulate fMRI and metabolomic data and analyze them
# by performing pre-processing and visualization. The main objective is to
# extract meaningful information about the brain connectivity patterns and 
# the molecular profiling of various diseases,
# let's say Alzheimer's or Schizophrenia.  

# The code was adapted from ChatGPT.
# For further information, you may visit https://openai.com/chatgpt

######################################################################

# First, we import the required modules/packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cross_decomposition import CCA
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

import networkx as nx

# The, we generate/simulate some fMRI and metabolomic data
def generate_data(number_subjects=200, number_regions=40, number_metabolites=15, disease_ratio=0.4):
    np.random.seed(35)

    labels = np.random.choice([0, 1], size=number_subjects, p=[1-disease_ratio, disease_ratio])

    Fmri_data = []
    Metabolomic_data = []

    for label in labels:
        # We create an fMRI connectivity matrix
        base_con = np.random.rand(number_regions, number_regions)
        con_matrix = (base_con + base_con.T) / 2

        # For the group speciric signal:
        if label == 1:  # state of disease
            con_matrix += np.eye(number_regions) * 0.2
        else:  # for the healthy state
            con_matrix += np.eye(number_regions) * 0.1

        Fmri_data.append(con_matrix[np.triu_indices(number_regions, k=1)])

# For the metabolomic data
        base_metabol = np.random.normal(0, 1, number_metabolites)
        if label == 1:
            base_metabol[:5] += 2  # let's upregulate some key metabolites
        Metabolomic_data.append(base_metabol)

    return np.array(Fmri_data), np.array(Metabolomic_data), labels

# Pre-processing the fMRI and metabolomic data:
def preprocess_data(Fmri_data, Metabolomic_data):
    scaler_Fmri = StandardScaler()
    scaler_Metabol = StandardScaler()
    Fmri_scaled = scaler_Fmri.fit_transform(Fmri_data)
    Metabol_scaled = scaler_Metabol.fit_transform(Metabolomic_data)
    return Fmri_scaled, Metabol_scaled

# Reducing the dimensionality of the fMRI and metabolomic data 
# (by performing principal component analysis - PCA)
def reduce_dimensions(Fmri_scaled, Metabol_scaled, number_components=8):
    PCA_Fmri = PCA(n_components=number_components)
    PCA_Metabol = PCA(n_components=number_components)
    Fmri_PCA = PCA_Fmri.fit_transform(Fmri_scaled)
    Metabol_PCA = PCA_Metabol.fit_transform(Metabol_scaled)
    return Fmri_PCA, Metabol_PCA, PCA_Fmri, PCA_Metabol

# Then, we perform canonical correlation analysis (CCA)
def run_cca(Fmri_PCA, Metabol_PCA, number_components=8):
    cca = CCA(n_components=number_components)
    cca_Fmri, cca_Metabol = cca.fit_transform(Fmri_PCA, Metabol_PCA)
    return cca_Fmri, cca_Metabol, cca

# Classifying fMRI and metabolomic data
def classify_multimodal_data(cca_Fmri, cca_Metabol, labels):
    Combined = np.hstack((cca_Fmri, cca_Metabol))
    X_train, X_test, y_train, y_test = train_test_split(Combined, labels, test_size=0.2, stratify=labels)

    clf = RandomForestClassifier()
    clf.fit(X_train, y_train)
    preds = clf.predict(X_test)
    print("Classification Report:\n", classification_report(y_test, preds))

# Visualizing the results
def plot_cca_scatter(cca_Fmri, cca_Metabol, labels):
    plt.figure(figsize=(10,5))
    plt.subplot(1,2,1)
    sns.scatterplot(x=cca_Fmri[:, 0], y=cca_Fmri[:, 1], hue=labels, palette='coolwarm')
    plt.title("Canonical Correlation Analysis - fMRI")
    plt.subplot(1,2,2)
    sns.scatterplot(x=cca_Metabol[:, 0], y=cca_Metabol[:, 1], hue=labels, palette='coolwarm')
    plt.title("Canonical Correlation Analysis - Metabolites")
    plt.tight_layout()
    plt.show()
    
# Creating a plot connectivity graph

def plot_connectivity_graph(average_con_vector, number_regions=40, title='Average Brain Connectivity'):
    con_matrix = np.zeros((number_regions, number_regions))
    iu = np.triu_indices(number_regions, k=1)
    con_matrix[iu] = average_con_vector
    con_matrix += con_matrix.T

    G = nx.from_numpy_array(con_matrix)
    plt.figure(figsize=(8, 6))
    pos = nx.spring_layout(G, seed=35)
    nx.draw(G, pos, node_size=50, edge_color='black', alpha=0.5)
    plt.title(title)
    plt.show()

# Running the whole pipeline
Fmri_data, Metabolomic_data, labels = generate_data()

Fmri_scaled, Metabol_scaled = preprocess_data(Fmri_data, Metabolomic_data)
Fmri_PCA, Metabol_PCA, _, _ = reduce_dimensions(Fmri_scaled, Metabol_scaled)
cca_Fmri, cca_Metabol, _ = run_cca(Fmri_PCA, Metabol_PCA)
classify_multimodal_data(cca_Fmri, cca_Metabol, labels)
plot_cca_scatter(cca_Fmri, cca_Metabol, labels)

# Visualizing the average connectivity between each group
healthy_average = Fmri_data[labels == 0].mean(axis=0)
diseased_average = Fmri_data[labels == 1].mean(axis=0)
plot_connectivity_graph(healthy_average, title="Connectivity in Health")
plot_connectivity_graph(diseased_average, title="Connectivity in Disease")

File: Multimodal_Analysis_Tsirakou.R
# The code below aims to simulate fMRI and metabolomic data and analyze them
# by performing pre-processing and visualization. The main objective is to
# extract meaningful information about the brain connectivity patterns and 
# the molecular profiling of various diseases,
# let's say Alzheimer's or Schizophrenia.  

# The code was adapted from ChatGPT.
# For further information, you may visit https://openai.com/chatgpt


# First, we install the required packages

install.packages("ggplot2")
install.packages("reshape")
install.packages("caret")
install.packages("randomForest")
install.packages("CCA")
install.packages("igraph")

# We then load the respective libraries

library(ggplot2)
library(reshape2)
library(caret)
library(randomForest)
library(CCA)
library(igraph)

# Step 1: We simulate some simulated fMRI and metabolomic data
data <- function(number_subjects = 300, number_regions = 50, number_metabolites = 20, disease_ratio = 0.5) {
set.seed(55) # we set a random seed for reproducibility
labels <- sample(c(0, 1), size = number_subjects, replace = TRUE, prob = c(1 - disease_ratio, disease_ratio))
  
upper_indices <- which(upper.tri(matrix(0, number_regions, number_regions)))
number_features <- length(upper_indices)
  
Fmri_data <- matrix(0, nrow = number_subjects, ncol = number_features)
Metabol_data <- matrix(0, nrow = number_subjects, ncol = number_metabolites)
  
  for (i in 1:number_subjects) {
    label <- labels[i]
    
    base_con <- matrix(runif(number_regions^2), nrow = number_regions)
    con_matrix <- (base_con + t(base_con)) / 2
    if (label == 1) {
      con_matrix <- con_matrix + diag(0.2, number_regions)
    } else {
      con_matrix <- con_matrix + diag(0.1, number_regions)
    }
    Fmri_data[i, ] <- con_matrix[upper.tri(con_matrix)]
    
    base_metabol <- rnorm(number_metabolites)
    if (label == 1) {
      base_metabol[1:5] <- base_metabol[1:5] + 2
    }
    Metabol_data[i, ] <- base_metabol
  }
  
  list(Fmri_data = Fmri_data, Metabol_data = Metabol_data, labels = labels)
}

# Step 2: Pre-processing of the simulated data
data_preprocessed <- function(Fmri_data, Metabol_data) {
Fmri_scaled <- scale(Fmri_data)
Metabol_scaled <- scale(Metabol_data)
list(Fmri_scaled = Fmri_scaled, Metabol_scaled = Metabol_scaled)
}

# Step 3: Dimensionality reduction by performing PCA
dimensions_reduced <- function(Fmri_scaled, Metabol_scaled, number_components = 10) {
Fmri_pca <- prcomp(Fmri_scaled)
Metabol_pca <- prcomp(Metabol_scaled)
list(Fmri_PCA = Fmri_pca$x[, 1:number_components], Metabol_PCA = Metabol_pca$x[, 1:number_components])
}

# Step 4: Canonical Correlation Analysis (CCA)
cca <- function(Fmri_PCA, Metabol_PCA, number_components = 10) {
cca_result <- cc(Fmri_PCA, Metabol_PCA)
cca_Fmri <- cca_result$scores$xscores[, 1:number_components]
cca_Metabol <- cca_result$scores$yscores[, 1:number_components]
list(cca_Fmri = cca_Fmri, cca_Metabol = cca_Metabol)
}

# Step 5: Performing classification using Random Forest
classify_multimodal_data <- function(cca_Fmri, cca_Metabol, labels) {
data_combined <- as.data.frame(cbind(cca_Fmri, cca_Metabol))
data_combined$label <- as.factor(labels)
  
set.seed(55)
train_index <- createDataPartition(data_combined$label, p = 0.8, list = FALSE)
train_data <- data_combined[train_index, ]
test_data <- data_combined[-train_index, ]
  
rf_model <- randomForest(label ~ ., data = train_data, ntree = 100)
preds <- predict(rf_model, test_data)
  
cat("Confusion Matrix:\n")
print(confusionMatrix(preds, test_data$label))
}

# Step 6: Scatter plots for Canonical Correlation Analysis
plot_cca_scatter <- function(cca_Fmri, cca_Metabol, labels) {
data_frame_1 <- data.frame(CC1 = cca_Fmri[, 1], CC2 = cca_Fmri[, 2], Label = as.factor(labels))
data_frame_2 <- data.frame(CC1 = cca_Metabol[, 1], CC2 = cca_Metabol[, 2], Label = as.factor(labels))
  
plot_1 <- ggplot(data_frame_1, aes(x = CC1, y = CC2, color = Label)) + 
    geom_point() + 
    ggtitle("CCA - fMRI")
  
plot_2 <- ggplot(data_frame_2, aes(x = CC1, y = CC2, color = Label)) + 
    geom_point() + 
    ggtitle("CCA - Metabolomics")
  
  gridExtra::grid.arrange(plot_1, plot_2, ncol = 2)
}

# 7. Plot connectivity graph
plot_connectivity_graph <- function(avg_vector, n_regions = 50, title_str = "Connectivity") {
con_matrix <- matrix(0, nrow = n_regions, ncol = n_regions)
con_matrix[upper.tri(con_matrix)] <- avg_vector
con_matrix <- con_matrix + t(con_matrix)
  
g <- graph_from_adjacency_matrix(con_matrix, mode = "undirected", weighted = TRUE, diag = FALSE)
plot(g, layout = layout_with_fr(g), vertex.size = 5, edge.color = "grey", main = title_str)
}

# Running the whole pipeline

# Step 1: Simulating the data
dataset <- data()
Fmri_data <- dataset$Fmri_data
Metabol_data <- dataset$Metabol_data
labels <- dataset$labels

# Step 2: Pre-processing
scaled <- data_preprocessed(Fmri_data, Metabol_data)

# Step 3: Principal Component Analysis (PCA)
pca_results <- dimensions_reduced(scaled$Fmri_scaled, scaled$Metabol_scaled)

# Step 4: CCA
cca_results <- cca(pca_results$Fmri_PCA, pca_results$Metabol_PCA)

# Step 5: Random Forest Classification
classify_multimodal_data(cca_results$cca_Fmri, cca_results$cca_Metabol, labels)

# Step 6: CCA scatter plots
plot_cca_scatter(cca_results$cca_Fmri, cca_results$cca_Metabol, labels)

# Step 7: Connectivity graphs
average_healthy <- colMeans(Fmri_data[labels == 0, ])
average_diseased <- colMeans(Fmri_data[labels == 1, ])
plot_connectivity_graph(average_healthy, title_str = "Connectivity - Healthy")
plot_connectivity_graph(average_diseased, title_str = "Connectivity - Diseased")
